designs/aws-launch-template-network-interfaces.md

```diff
@@ -0,0 +1,126 @@
+  amiSelector: { ... }                    # optional, discovers tagged amis to override the amiFamily's default
+  userData: "..."                         # optional, overrides autogenerated userdata with a merge semantic
+  tags: { ... }                           # optional, propagates tags to underlying EC2 resources
+  metadataOptions: { ... }                # optional, configures IMDS for the instance
+  blockDeviceMappings: [ ... ]            # optional, configures storage devices for the instance
+  detailedMonitoring: "..."               # optional, configures detailed monitoring for the instance
+  networkInterfaces:                      # optional, configures network interfaces for the instance
+    - associateCarrierIPAddress: false    # optional, Indicates whether to associate a Carrier IP address with eth0 for a new network interface.
+      associatePublicIPAddress: false     # optional, Indicates whether to assign a public IPv4 address to eth0 for a new network interface.
+      deleteOnTermination: false          # optional, Indicates whether the network interface is deleted when the instance is terminated.
+      description: "..."                  # optional, A description for the network interface.
+      deviceIndex: 0                      # required, The device index for the network interface attachment.
+      interfaceType: "interface"          # optional, The interface type for the network interface. To create an Elastic Fabric Adapter (EFA), specify efa
+      networkCardIndex: 0 
+      ipv4PrefixCount: 1                  # optional, The number of IPv4 prefixes that AWS automatically assigns to the network interface.
+      ipv6PrefixCount: 1                  # optional, The number of IPv6 prefixes that AWS automatically assigns to the network interface.
+    - { ... }
+```
+
+subnet and securityGroup will remain part of the `AWSNodeTemplate` and will be used for both network interfaces.
+in following iterations, as the `AWSNodeTemplate` move from `v1alpha1` to higher versions, we can add support for:
+
+```yaml
+subnetSelector: String
+securityGroupSelector: String
+```
+
+this way we can support different subnets and security groups for each network interface, we can also remove the limit
+of having only 2 network interfaces per LaunchTemplate.
+
+The [drift design doc #366](https://github.com/aws/karpenter-core/pull/366) suggests classifying CRD fields as  (1) Static, (2) Dynamic, or (3) Behavioral. in this
+context, the networkInterfaces would be classified as following:
+
+|                                                                | Static | Dynamic | Behavioral |
+|----------------------------------------------------------------|--------|---------|------------|
+| AWSNodeTemplate.NetworkInterfaces.length                       | x      |         |            |
+| AWSNodeTemplate.NetworkInterfaces.order                        | x      |         |            |
```

Comment:
Does the order in which network interfaces are defined define some priority on how they're being used? If so, then this makes sense to me, but otherwise, I don't know if we need to consider the order.

Reply:
Yea, the order does not change the way they are used. It is the device index that matters.

after reconsiderations, the order in which they are defined does not matter, we should only compare the network interfaces where the device index match

I think in this case it is better to add the validation to have unique value per device index in the `AWSNodeTemplate` (already enforced by AWS API but better to have the AWSNodeTemplate in a valid state as well)